Drain the queue asynchronously if it's overloaded with redundant items.

sujithvaddi opened the pull request:
## Github Issue #

[1234](https://github.com/bazaarvoice/emodb/issues/11)

## What Are We Doing Here?

Recently we encountered an issue where a client wrote a flood up non-mutative updates to Emo. As a result most databus subscriptions grew to contain millions of raw events. In such cases, we would like to drain the queue asynchronously to delete the redundant events from the queue. 

## How to Test and Verify

1. Check out this PR
2. Create a table and with a document. 
3. Create a databus subscription and then add non-mutative deltas to the table.
4. Poll the subscription for non-mutative deltas, verify the draining task has kicked in from the logs and also check the size of the queue to confirm if all are deleted.

## Risk

### Level 

`Low`, `Medium`, or `High`. Give an indication of what you think is the level of change introduced by this PR. `High` means a massive change to a core functionality. 
`Low` means a really minor change that shouldn't have any regression effect. 

### Required Testing

`Smoke`, `Regression`, or `Manual`. (All changes except documentation need smoke
testing at a minimum).

### Risk Summary

Add one or a few complete sentences about the possible risks or concerns for
this change.

## Code Review Checklist

- [ ] Tests are included. If not, make sure you leave us a line or two for the reason.

- [ ] Pulled down the PR and performed verification of at least being able to
build and run.

- [ ] Well documented, including updates to any necessary markdown files. When
we inevitably come back to this code it will only take hours to figure out, not
days.

- [ ] Consistent/Clear/Thoughtful? We are better with this code. We also aren't
a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.

- [ ] PR has a valid summary, and a good description.

billkalter commented on databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java:
```diff
@@ -740,7 +790,109 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
+        boolean nonRedundantItemFound = false;
+
+        // this loop will break once the peek found a non-redundant item or when the queue gets empty.
+        while (true) {
```
Instead of using a while loop I'd recommend re-submitting the task back onto `_drainService`.  As currently written once there are 10 subscriptions in need of draining the 11th onward are blocked until one of the others finish.  By re-submitting on each iteration it provides some level of fairness, or at least minimizes the chances of starvation.

sujithvaddi commented:
@billkalter addressed the feedback. 

Did manual testing and verified. I'll be adding a unit test if possible.

Also, for the ownership thing, I'm using the Time instead for now.
 _I tried to see if I can extend the interface by adding a new method to EventStore code which will use the OstrictOwnerGroup’s isOwner() method. But the existing OwnerGroup seemed to be only relevant to DedupEventStore context and not with the DefaultEventStore, so it’s not straight forward to expose it to DatabusEventStore unless I skip reusing the existing ownerGroup code and do other reimplementations._

billkalter commented on databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java:
```diff
@@ -740,7 +780,111 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
+                _log.info("Draining for subscription: {} was already started from a previous poll.", subscription);
+            }
+        } catch (Exception e) {
+            _log.error("Encountered exception while draining the queue for subscription: {}.", subscription);
```
Add `e` as the final parameter on `_log.error()` so the root cause gets logged.

billkalter commented on databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java:
```diff
@@ -740,7 +780,111 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
+                // Is the change redundant?
+                if (readResult.isChangeDeltaRedundant(changeId)) {
+                    eventIdsToDiscard.add(eventId);
+                    continue;
```
Nit-pick:  There's only one line after this `if` block.  Instead of a `continue` why not just put the following line in an `else` clause?

billkalter commented on databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java:
```diff
@@ -740,7 +780,111 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
+            _eventStore.delete(subscription, eventIdsToDiscard, true);
+        }
+
+        long totalSubscriptionDrainTime = _drainedSubscriptionsMap.get(subscription).longValue() + stopwatch.elapsed(TimeUnit.MILLISECONDS);
```
Do we want 1 minute total time draining the queue or 1 minute wall clock time since we first started draining?

sujithvaddi commented:
@billkalter Added tests, and addrssed feedback.

billkalter commented on databus/src/test/java/com/bazaarvoice/emodb/databus/core/DefaultDatabusTest.java:
```diff
@@ -61,4 +84,167 @@ public void testSubscriptionCreation() {
+            }
+        };
+        Map<String, Object> content = entity("table", "key", ImmutableMap.of("rating", "5"));
+        // Create a custom annotated content which returns all changes as redundant
```
Minor:  Comment should read "... as _not_ redundant"

billkalter commented on databus/src/main/java/com/bazaarvoice/emodb/databus/core/DefaultDatabus.java:
```diff
@@ -751,7 +795,112 @@ private void checkSubscriptionOwner(String ownerId, OwnedSubscription subscripti
+
+        // delete the events we never again want to see.
+        if (!eventIdsToDiscard.isEmpty()) {
+            _eventStore.delete(subscription, eventIdsToDiscard, true);
```
Minor suggestion:  Maybe put a metered metric around async redundant items drained, just to give some visibility into when this happens.

billkalter commented:
Note:  Failure of gatekeeper build is due to an unrelated change which requires a gatekeeper pull request to be merged.

sujithvaddi commented:
@billkalter added few metrics.

billkalter commented:
@sujithvaddi 👍
